teps/0048-pipeline-as-code.md

```diff
@@ -0,0 +1,214 @@
+- The source code is automatically tested with the pipeline definition in `.tekton`.
+
+Subsequently if the user needs to add another component to its source code (i.e:
+a javascript based frontend) and need to make a changes to the pipeline
+definition to do the testing of that new component, the change to the pipeline
+would only apply for that Pull Request until it will be merged.
+
+## Requirements
+
+* We need to be able to test changes inside the `.tekton` folder as part of the PR sent.
+* We should provide an easy way (single curl command and/or operator flag) to deploy Tekton with "tekton-as-code" support.
+* We need assume single tenancy on the pipelines-as-code cluster.
+* We assume single tenancy on the tekton-as-code cluster
+* We assume the user does not need access to the cluster, 
+
+## Proposal
+
+Create an initial set of triggers templates for the users to install on its cluster.
+Create an initial pipeline to test the change of a PR according to the templates in the `.tekton` project directory.
+
+Provide an one-command way to install a version for code with "Tekton as code"
+enabled that outputs a webhook URL and a secret that can be used by users to
+enable Tekton as code for their repo, and a readonly tekton dashboard that can
+be used to observe their pipeline runs.
+
+### Notes/Caveats
+
+With this initial proposal we have left out the notification part of Pipeline as
+a code. To make it a 'complete package' we would want Pipeline as Code being
+able to notify on PR the status of the `PipelineRun` execution, have the ability
+to see the live execution of the Pipeline or when the pipeline finished to know
+if it was succesfull or failed.
+
+With this initial proposal we have focused on a PR based workflow and not on the
+post-merge workflows. But there is nothing in this design that can't
+address post-merge workflows.
+
+We are not addressing here on how to check out private repositories and how to
+get token from the Web VCS provider (ie: via Github or Gitlab [personal
+token](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/creating-a-personal-access-token)
+or via an [Oauth
+application](https://docs.github.com/en/free-pro-team@latest/developers/apps))
+
+### Risks and Mitigations
+
+
+#### Logs retention
+
+There is currently no way with Tekton to have the logs in different locations
+which means we assume the logs will be kept as long as the `Pipelinerun` is
+still in the cluster.
+
+### User Experience
+
+* A user deploy with a single command the pipelines as code template and retrieve
+  the ingress url of the eventlistenner webhook url.
+* The user configure the webhook URL in GitHub or other web vcs provider.
+* A user add a tekton template in the `.tekton` directory of its repository.
+* User sends a PR.
+* Pipeline as code start applying the template located in the `.tekton`
+  directory
+* User is able to watch the pipeline execution status via `tkn` or `dashboard`.
```

Comment:
I'm curious how the authors visualize this happening on the dashboard. Right now there's not a lot of filters other than namespaces and labels. Are these filters going to be used to find different variants of same pipeline in the same UI or would this involve some restructuring of the UI as well?

Reply:
We chatted among RHT colleagues on how we wanted to present this if we wanted to expose pipeline as a code on the openshift console, which probably would be a bit the same for dashboard.

I would imagine, there would be a repository first view, where each repository can have a pipeline as a code status, with the different runs and link to logs.

Not so different from what we have on github actions and other CI i guess.

Reply:
it would be a bit of a leap for the tekton dashboard to become aware of repositories - but it would be nice! reminds me of https://github.com/tektoncd/dashboard/pull/1612 which is suggesting a grouping mechanism that could group triggering info + pipelines - i think @dibyom might be planning to look into proposing a grouping mechanism like this